### ISSUE TITLE
[Bug] HMR does not work for raw imported html

### ISSUE THREAD

**Issue by niuiic** [farm-fe/farm#774]
### Steps to reproduce

```
git clone https://github.com/niuiic/gallery
pnpm i
pnpm build
pnpm dev --filter css-components
```

Open the console.

Edit `packages/css/src/components/wavyUnderline/wavyUnderline.html`, the output does not change.
![image](https://github.com/farm-fe/farm/assets/63799565/38a6b4c8-dd45-41d9-8951-89771e475f81)

Edit `packages/css/src/components/wavyUnderline/wavyUnderline.scss`, the output changes.

`packages/css/src/components/wavyUnderline/index.ts`
```typescript
import html from './wavyUnderline.html?raw'
import css from './wavyUnderline.scss'
import cssStr from './wavyUnderline.scss?raw'

console.log(html)
console.log(cssStr)

export { css, html }
```

Another issue.
`packages/farm-plugins/src/raw.ts`, this farm plugin does not work properly. There may be conflicts with the builtin raw import feature. Can this feature be overrided by setting `priority`?

If it works, raw import should be transformed to (export default `hello`).

Cli output
![image](https://github.com/farm-fe/farm/assets/63799565/cd496e16-c613-4361-8011-06a1487c8086)

1. `transform` was triggered and returned correct value, but `cssStr` imported from `./wavyUnderline.scss?raw` is `.text ...`.
2. Only scss files were resolved, html files can only sometimes be resolved.

```typescript
import type { DevServer, JsPlugin, UserConfig } from '@farmfe/core'
import { readFileSync } from 'fs'

export const farmRawPlugin = (): JsPlugin => {
  let farmConfig: UserConfig['compilation']
  let devServer: DevServer

  return {
    name: 'farm-raw-plugin',
    priority: 10000,
    config: (args) => (farmConfig = args),
    configDevServer: (args) => (devServer = args),
    load: {
      filters: {
        resolvedPaths: ['\\?raw']
      },
      executor: (args) => {
        console.log(args.resolvedPath)
        return { content: readFileSync(args.resolvedPath, 'utf8'), moduleType: 'raw' }
      }
    },
    transform: {
      filters: {
        moduleTypes: ['raw']
      },
      executor: (args) => {
        console.log(args.resolvedPath)
        if (farmConfig?.mode === 'development') {
          devServer?.addWatchFile(args.resolvedPath, [])
        }

        console.log('export default ' + '`' + 'hello' + '`')
        return { content: 'export default ' + '`' + 'hello' + '`', moduleType: 'js' }
        // return { content: 'export default ' + '`' + args.content + '`', moduleType: 'js' }
      }
    }
  }
}
```

### What is expected?

The output of `console.log(html)` changes.

### What is actually happening?

The output of `console.log(html)` does not change.

### System Info

```Shell
System:
    OS: Linux 6.1 Gentoo Linux
    CPU: (16) x64 AMD Ryzen 7 5800H with Radeon Graphics
    Memory: 12.81 GB / 15.59 GB
    Container: Yes
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.1.0 - /run/user/1000/fnm_multishells/118331_1700722519002/bin/node
    npm: 10.2.0 - /run/user/1000/fnm_multishells/118331_1700722519002/bin/npm
    pnpm: 8.10.5 - /run/user/1000/fnm_multishells/118331_1700722519002/bin/pnpm
  npmPackages:
    @farmfe/core: ^0.14.4 => 0.14.4
```
```


### Any additional comments?

_No response_

**Comment by wre232114** [comment]
you can set priority to `-1000` to execute your plugins at last

**Comment by wre232114** [comment]
for load hook, you need to set larger priority to make sure your plugin execute first. but for transform hook, you may need to set smaller priority to make sure your plugins execute after all other transformers

**Comment by niuiic** [comment]
Not work.
![image](https://github.com/farm-fe/farm/assets/63799565/04342c27-e9ec-47c5-8019-c8ca7edc4249)

This makes me a little confused.

In my understanding, in general, If my plugin runs earlier, and return a no-null value, then other plugins should be skipped.

If, as you say, I want `load` to execute as early as possible and `transform` to execute as late as possible, I have to split the implementation into two plugins.

**Comment by wre232114** [comment]
load and transform are different kind of hooks. load is a `first hook` means other plugins will be skipped when a plugin return non-null value. but transform is a `serial hook` means every plugins' transform hook will be executed and the previous result will be passed to the next.

you need to implement 2 plugins, you can return a array that contains two plugin object.

I'll take a look at hmr of html raw imported

**Comment by wre232114** [comment]
The conflicts of builtin raw should be solved too. I'll fix it, when a user plugin returns result in transform, builtin plugins should ingore raw

**Comment by niuiic** [comment]
HMR problem has not been fixed.

Neither my plugin nor the built-in feature works.

BTW, it seems that sometimes farm caches the plugins. When I rebuild the plugin, I have to remove `node_modules/.farm` to make changes work.

@wre232114

**Comment by wre232114** [comment]
you can set `persistentCache` to `false` to disable cache when build your plugins. I tested your repo yestoday and it works.

**Comment by niuiic** [comment]
> you can set `persistentCache` to `false` to disable cache when build your plugins. I tested your repo yestoday and it works.

Got it. it's not a bug.

**Comment by niuiic** [comment]
![show](https://github.com/farm-fe/farm/assets/63799565/ea6e5c1c-d83d-4227-87ad-ee149e3b6350)

**Comment by wre232114** [comment]
i tested, HMR did not work with builtin raw plugin. but it works with your raw plugin

**Comment by wre232114** [comment]
I think the reason is that builtin raw plugin does not add the file to watch list, but your raw plugin does, that's a bug with the builtin feature

**Comment by niuiic** [comment]
Emm, it does not work.

I tried these two versions of my plugin.

```
import type { DevServer, JsPlugin, UserConfig } from '@farmfe/core'
import { readFileSync } from 'fs'

// export const farmRawPlugin = (): JsPlugin[] => {
//   let farmConfig: UserConfig['compilation']
//   let devServer: DevServer

//   return [
//     {
//       name: 'farm-raw-plugin-load',
//       priority: 1000,
//       load: {
//         filters: {
//           resolvedPaths: ['\\?raw']
//         },
//         executor: (args) => ({ content: readFileSync(args.resolvedPath, 'utf8'), moduleType: 'raw' })
//       }
//     },
//     {
//       name: 'farm-raw-plugin-transform',
//       priority: -1000,
//       config: (args) => (farmConfig = args),
//       configDevServer: (args) => (devServer = args),
//       transform: {
//         filters: {
//           moduleTypes: ['raw']
//         },
//         executor: (args) => {
//           if (farmConfig?.mode === 'development') {
//             devServer?.addWatchFile(args.resolvedPath, [])
//           }

//           return { content: 'export default ' + '`' + args.content + '`', moduleType: 'js' }
//         }
//       }
//     }
//   ]
// }

export const farmRawPlugin = (): JsPlugin => {
  let farmConfig: UserConfig['compilation']
  let devServer: DevServer

  return {
    name: 'farm-raw-plugin',
    priority: 1000,
    config: (args) => (farmConfig = args),
    configDevServer: (args) => (devServer = args),
    load: {
      filters: {
        resolvedPaths: ['\\?raw']
      },
      executor: (args) => ({ content: readFileSync(args.resolvedPath, 'utf8'), moduleType: 'raw' })
    },
    transform: {
      filters: {
        moduleTypes: ['raw']
      },
      executor: (args) => {
        if (farmConfig?.mode === 'development') {
          devServer?.addWatchFile(args.resolvedPath, [])
        }

        return { content: 'export default ' + '`' + args.content + '`', moduleType: 'js' }
      }
    }
  }
}
```

**Comment by niuiic** [comment]
![image](https://github.com/farm-fe/farm/assets/63799565/b1bb9526-a8a5-4bbb-9380-0d585c091052)

![image](https://github.com/farm-fe/farm/assets/63799565/6c11603c-c721-4a31-b34e-1e65bdafabe6)

Only the scss file changes were detected.

**Comment by wre232114** [comment]
```ts
export const farmRawPlugin = (): JsPlugin => {
  let farmConfig: UserConfig['compilation']
  let devServer: DevServer

  return {
    name: 'farm-raw-plugin',
    priority: 1000,
    config: (args) => (farmConfig = args),
    configDevServer: (args) => (devServer = args),
    load: {
      filters: {
        resolvedPaths: ['\\?raw']
      },
      executor: (args) => ({ content: readFileSync(args.resolvedPath, 'utf8'), moduleType: 'raw' })
    },
    transform: {
      filters: {
        moduleTypes: ['raw']
      },
      executor: (args) => {
        if (farmConfig?.mode === 'development') {
          devServer?.addWatchFile(args.resolvedPath, [])
        }

        return { content: 'export default ' + '`' + args.content + '`', moduleType: 'js' }
      }
    }
  }
}
```

I tried this version works

**Comment by niuiic** [comment]
It's same to the second version I tried.

**Comment by wre232114** [comment]
I just copied it from your comments. try remove cache of css components

**Comment by wre232114** [comment]
or try add your plugins path to `persistentCache.buildDependencies` to invalidate cache when your plugins changed

**Comment by wre232114** [comment]
Or cache of turbo may have impact, try `run build` under `farm-plugins` for a fresh build

**Comment by wre232114** [comment]
I confirmed that it's a bug caused by farm cache. it works when cache does not exist, but it does not work when start twice with cache

**Comment by niuiic** [comment]
I reclone the repo. And it never works. It does not work when there is no cache.

**Comment by wre232114** [comment]
> ```ts
> export const farmRawPlugin = (): JsPlugin => {
>   let farmConfig: UserConfig['compilation']
>   let devServer: DevServer
> 
>   return {
>     name: 'farm-raw-plugin',
>     priority: 1000,
>     config: (args) => (farmConfig = args),
>     configDevServer: (args) => (devServer = args),
>     load: {
>       filters: {
>         resolvedPaths: ['\\?raw']
>       },
>       executor: (args) => ({ content: readFileSync(args.resolvedPath, 'utf8'), moduleType: 'raw' })
>     },
>     transform: {
>       filters: {
>         moduleTypes: ['raw']
>       },
>       executor: (args) => {
>         if (farmConfig?.mode === 'development') {
>           devServer?.addWatchFile(args.resolvedPath, [])
>         }
> 
>         return { content: 'export default ' + '`' + args.content + '`', moduleType: 'js' }
>       }
>     }
>   }
> }
> ```
> 
> I tried this version works

reclone the repo and use this version

**Comment by wre232114** [comment]
<img width="724" alt="image" src="https://github.com/farm-fe/farm/assets/8372439/074b4af9-42d0-4186-8438-55f0993722cc">

**Comment by niuiic** [comment]
Reclone the repo, use this version, build without turbo, but no luck.

**Comment by wre232114** [comment]
maybe, there does have a bug, it should work without any custom plugins, i'll fix it

**Comment by wre232114** [comment]
fixed in 0.14.6

**Comment by niuiic** [comment]
It is fixed.
